src/tvheadend/utilities/SettingsMigration.cpp

```diff
@@ -0,0 +1,127 @@
+
+namespace
+{
+// <setting name, default value> maps
+const std::vector<std::pair<const char*, const char*>> stringMap = {
+    {"host", "127.0.0.1"}, {"user", ""}, {"pass", ""}, {"wol_mac", ""}, {"streaming_profile", ""}};
+
+const std::vector<std::pair<const char*, int>> intMap = {{"htsp_port", 9981},
+                                                         {"http_port", 9982},
+                                                         {"connect_timeout", 10000},
+                                                         {"response_timeout", 5000},
+                                                         {"total_tuners", 1},
+                                                         {"pretuner_closedelay", 10},
+                                                         {"autorec_approxtime", 0},
+                                                         {"autorec_maxdiff", 15},
+                                                         {"dvr_priority", DVR_PRIO_NORMAL},
+                                                         {"dvr_lifetime2", 15},
+                                                         {"dvr_dubdetect", DVR_AUTOREC_RECORD_ALL},
+                                                         {"stream_readchunksize", 64}};
+
+const std::vector<std::pair<const char*, bool>> boolMap = {{"https", false},
+                                                           {"epg_async", true},
+                                                           {"pretuner_enabled", false},
+                                                           {"autorec_use_regex", false},
+                                                           {"streaming_http", false},
+                                                           {"dvr_playstatus", true},
+                                                           {"dvr_ignore_duplicates", true}};
+
+} // unnamed namespace
+
+bool SettingsMigration::MigrateSettings(kodi::addon::IAddonInstance& target)
+{
+  std::string stringValue;
+  bool boolValue{false};
+  int intValue{0};
+
+  if (target.CheckInstanceSettingString("kodi_addon_instance_name", stringValue) &&
+      !stringValue.empty())
+  {
+    // Instance already has valid instance settings
+    return false;
+  }
+
+  // Read pre-multi-instance settings from settings.xml, transfer to instance settings
+  SettingsMigration mig(target);
+
+  for (const auto& setting : stringMap)
+    mig.MigrateStringSetting(setting.first, setting.second);
+
+  for (const auto& setting : intMap)
+    mig.MigrateIntSetting(setting.first, setting.second);
+
+  for (const auto& setting : boolMap)
+    mig.MigrateBoolSetting(setting.first, setting.second);
+
+  if (mig.Changed())
+  {
+    // Set a title for the new instance settings
+    std::string title;
+    target.CheckInstanceSettingString("host", title);
+    if (title.empty())
+      title = "Migrated Add-on Config";
+
+    target.SetInstanceSettingString("kodi_addon_instance_name", title);
```

Comment:
Would perhaps be better to use a dedicated setting for this, e.g. `setting_migration_done`, that way we don't run into unexpected issues if for whatever reason the `kodi_addon_instance_name` setting is later renamed or removed.

Reply:
Yes, definitely better. I will change that.

Reply:
On second thought, I think we actually don't need another setting. The only thing that might go wrong is that if somebody removes the instance title by hand from intsance-settings-1.xml (Kodi GUI prevents this), we migrate the settings for this instance a second time, overwriting the previously user-set values with the old values from settings.xml. Not nice, but if users manipulate files in a way they are not supposed to to, well ...

Reply:
Alright, sure!